helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -490,66 +495,75 @@ private boolean updateOnHostChange(final DeploymentGroup group,
-        curHostsVersion = Optional.of(chn.getStat().getVersion());
-        curHosts = Json.read(chn.getBytes(), new TypeReference<List<String>>() {});
-      } catch (JsonMappingException e) {
-        curHosts = Collections.emptyList();
-      }
-
       final DeploymentGroupStatus status = getDeploymentGroupStatus(groupName);
       if (!allowHostChange(status)) {
         return;
       }
-      if (!curHostsVersion.isPresent() || !hosts.equals(curHosts)) {
-        // Node not present or hosts have changed
-        final List<ZooKeeperOperation> ops = Lists.newArrayList();
-        ops.add(set(Paths.statusDeploymentGroupHosts(groupName), Json.asBytes(hosts)));
-
-        final Node dgn = client.getNode(Paths.configDeploymentGroup(groupName));
-        final Integer deploymentGroupVersion = dgn.getStat().getVersion();
-        DeploymentGroup deploymentGroup = Json.read(dgn.getBytes(), DeploymentGroup.class);
-
-        ImmutableList<Map<String, Object>> events = ImmutableList.of();
-
-        if (deploymentGroup.getJobId() != null) {
-          if (updateOnHostChange(deploymentGroup, status)) {
-            deploymentGroup = deploymentGroup.toBuilder()
-                .setRollingUpdateReason(HOSTS_CHANGED)
-                .build();
-            // Fail transaction if the deployment group has been updated elsewhere.
-            ops.add(check(Paths.configDeploymentGroup(groupName), deploymentGroupVersion));
-
-            // NOTE: If the DG was removed this set() cause the transaction to fail, because
-            // removing the DG removes this node. It's *important* that there's an operation that
-            // causes the transaction to fail if the DG was removed or we'll end up with
-            // inconsistent state.
-            ops.add(set(Paths.configDeploymentGroup(groupName), deploymentGroup));
-
-            final RollingUpdateOp op = getInitRollingUpdateOps(deploymentGroup, hosts, client);
-            ops.addAll(op.operations());
-            events = op.events();
-          }
-        }
 
-        log.info("starting zookeeper transaction for updateDeploymentGroupHosts on "
-                 + "deployment-group name {} jobId={}. List of operations: {}",
-            groupName, deploymentGroup.getJobId(), ops);
+      final List<String> curHosts = getHosts(client, Paths.statusDeploymentGroupHosts(groupName));
+      final List<String> previouslyRemovedHosts = getHosts(
+          client, Paths.statusDeploymentGroupRemovedHosts(groupName));
+
+      if (hosts.equals(curHosts)) {
+        return;
+      }
+
+      final List<String> removedHosts = removedHosts(curHosts, hosts, previouslyRemovedHosts);
+
+      final List<ZooKeeperOperation> ops = Lists.newArrayList();
+      ops.add(set(Paths.statusDeploymentGroupHosts(groupName), Json.asBytes(hosts)));
+      ops.add(set(Paths.statusDeploymentGroupRemovedHosts(groupName), Json.asBytes(removedHosts)));
```

Comment:
> does the list of hosts stored in this path for "removed hosts" grow unbounded over time? 

@mattnworb Hosts are removed from this list if and when they are 'marked undeployed' per https://github.com/spotify/helios/pull/986/commits/58fedba3cae84ed8125fee61827057d2d9cf26b7#diff-8d500c531ffb1f7836e9d5b60585a002R979.

I *think* hosts will be 'marked undeployed' if they're unregistered from the Helios master, because the `getDeployment(host, jobid)` check at  https://github.com/spotify/helios/pull/986/commits/58fedba3cae84ed8125fee61827057d2d9cf26b7#diff-8d500c531ffb1f7836e9d5b60585a002R948 will come up blank and assume the job is gone, causing the 'await stopped' rollout task to proceed to the 'mark undeployed' task.

However any agents that are removed *without* unregistering will cause the list of stored hosts to grow unbounded, and potentially worse will cause any rolling update of the deployment groups they match to fail until the agent either comes back online or is unregistered.